### ISSUE TITLE
Feedback: README

### ISSUE THREAD

**Issue by tooolbox** [switchupcb/copygen#16]
I consider myself to be an experienced Go developer and have used various codegen tools in the past including SQLBoiler.  I've been grappling with the issue of mapping data between different representations (structs) at different levels of an application, and came across both goverter and copygen.

Copygen seems like it's solving the problem I want to solve, the way I want to solve it: CLI tool to generate zero-overhead mapping code between structs.  Therefore I'm quite interested.

I read through the README and I have some points of feedback and questions.  They may or may not be palatable and I acknowledge I'm spouting off some opinions 10 minutes after finding your project, but first impressions from would-be users can have a certain value so please take it in that light.

1. Looking at the README overall, I think a more simple onramp would be valuable.  If you look at [goverter](https://github.com/jmattheis/goverter) it starts with a straightforward 1-to-1 mapping with identical field names and types, and the examples get more complex from there.  In the copygen README, you're starting off with a 2-to-1 struct mapping, so off the bat I'm spending brain cycles trying to figure out how the made-up model objects are supposed to relate to one another, rather than just grokking the tool.  I suggest cutting back the gradient so you start with an absolute rock-bottom example and take it up one piece at a time.
2. Moving on, the README states that setup involves a `.go` file and a `.yml` file.  That seems like perhaps excessive moving parts and setup for the most basic scenarios.  The most vital component of the `.yml` file seems to be specifying input and output file names, but sane defaults could be used to allow things to "just work" with only the `.go` file.  Perhaps it looks for `copygen.go` and `copygen.yml` by default, and generates `copygen.gen.go`?  Thoughts?
3. Moving on to the `setup.go` section, it makes sense until the statement `Copygen uses no allocation with pointers which means fields are assigned to objects passed as parameters.`  This is confusing because the signature you show just above it is `ModelsToDomain(models.Account, models.User) *domain.Account` which I gather is creating a `domain.Account` and returning a pointer to it; it shouldn't be assigning to the objects that are passed as parameters.  I am probably misunderstanding; a clarification or better wording would be valuable.
4. On the `options` header:
    - I would suggest changing away from a table to a format that doesn't introduce line breaks, since said line breaks make it more difficult to visually parse the examples.  
    - Suggest always showing fully concrete examples before showing examples that use regexes.  
    - Instead of just saying `map from to` I would suggest `map <from> <to>` or similar, so it's easy to see what's literal and what's meant to be variable.
    - Suggest showing actual code examples for each of these so you can see how they work in practice.
5. In the `output` of the first example, it seems like it copied the Itoa function.  If I already wrote the function in one `.go` file, it doesn't seem like I'd need it defined again.  I suppose if you're using a `setup.go` that's in one directory and putting your `copygen.go` in another directory, yes you'd need to move the bits over.  But why not put it all in the same place and avoid that complexity?
6. The `Customization` section is, at the risk of sounding whiny and negative, a little mind-boggling.  I don't understand the use cases behind Custom Types or Templates.  Okay, I see that you need to use Templates to add an error-return although that seems a little extreme, I believe such is available out of the box with goverter.
7. Rather than jumping into deep customization, let's say I want to do a 1-to-1 struct mapping where a couple of the fields have different names and a couple have different types (int vs. int64 or int vs. string, let's say).  I want to leverage copygen's automatic mapping for the most part and just provide a couple options, arguments, utility functions, what have you, to deal with the special cases.  An example of this would be awesome.
8. There's not really enough data given under Matcher, apparently there's manual methods and there's automatic methods.  Examples and comparisons would be great.
9. The depth example is hard to parse, I see that it's supposed to illustrate field depth but the indentation is a little confusing.  I would suggest some curly braces so you get better delineation of where the sub-structs start and end.  And then, actual code examples that perform mapping using the detailed example of depth that was just provided.

Apologies if any of this comes across as harsh or entitled, it's not meant that way.  I'm very interested in using a tool like copygen for my projects, and this one seems overall more powerful and flexible than goverter, but I find myself right away bouncing off of some confusions and perceived complexity, and I wanted to give some feedback on this.

**Comment by switchupcb** [comment]
@tooolbox 
I appreciate the feedback. If you haven't noticed already, examples with in-depth explanation are provided from [#usage](https://github.com/switchupcb/copygen#Usage) or by checking the [`examples` ](https://github.com/switchupcb/copygen/tree/main/examples) folder.

> 1.
A 2-1 structure is used because it is the base scenario.  There is a domain model with a struct `ID`, `UserID`, `Name`. There are 2 database models with an Account struct containing `ID`, `Name`, and User struct containing `UserID`; its `Name` isn't copied. The example takes the 3 necessary data fields and copies them into the domain fields. **Please explain how this can be grokked better.**

Would copygen apply to a 1:1 map? In goverter's example, it seems redundant to have an extra struct when you already have a struct with the necessary fields.

**Comment by switchupcb** [comment]
> 5.
`setup.go` is the UI, but not part of the project. No code you write there is actually used by the output you import into other files; because it's not imported.

**Comment by switchupcb** [comment]
> 6. 
If you can't see a use-case, that doesn't mean there isn't one. If you can return errors using goverter, that's all you can do. Copygen has more features and more customization options.

**Comment by switchupcb** [comment]
> 4. See examples and/or provide example of format change.
> 7. See examples and/or provide an example using the format from examples.
> 8. See examples.
> 9. See examples.

See examples from https://github.com/switchupcb/copygen/issues/16#issuecomment-1031715716.

**Comment by tooolbox** [comment]
> Would copygen apply to a 1:1 map? In goverter's example, it seems redundant to have an extra struct when you already have a struct with the necessary fields. Let's also clarify that a 1:1 map is possible. However, it seems redundant once you know how to use the first example or any other example.

The use case of 1:1 is when you are mapping between structs in different layers that have similar but not necessarily identical shapes.  It seems like the best "base scenario" for someone new to your library.  A 2:1 example is by definition not as much of a "base scenario".

> I can see how this can be misunderstood. Please let me know if this change would fix it.

Not really.  This is effectively what your sentence says:

> Copygen uses no allocation with pointers, and as a consequence of that, fields are assigned to objects passed as parameters.

This doesn't make sense because in your example of `ModelsToDomain(models.Account, models.User) *domain.Account` you're not assigning to anything passed as a parameter.  In that example, all arguments are by-value, not by-reference, and all data is assigned to a newly created struct, and you return a pointer to that newly-created struct.  At least, that's what your example looks like.  So then you follow it up with an explanation about assigning to the parameters and it's confusing and directly contradictory.

------------------------------------------

Generally, you seem more defensive than interested in changing anything about your docs to make your project more approachable.  That's understandable, since I'm a random stranger coming and criticizing your work.  Apologies, and I'll move on.

**Comment by switchupcb** [comment]
@tooolbox
> Generally, you seem more defensive than interested...

Not sure why you think this.

> The use case of 1:1 is when you are mapping between structs in different layers that have similar but not necessarily identical shapes. It seems like the best "base scenario" for someone new to your library. A 2:1 example is by definition not as much of a "base scenario".

The only difference is using a 1:1 map which is using 2 structs instead of three in the parameter list. Have you read any of the examples yet?

> Copygen uses no allocation with pointers, and as a consequence of that, fields are assigned to objects passed as parameters.

This is correct.

> This doesn't make sense because in your example of ModelsToDomain(models.Account, models.User) *domain.Account you're not assigning to anything passed as a parameter...
```go
// ModelsToDomain copies a Account, User to a Account.
func ModelsToDomain(tA *domain.Account, fA models.Account, fU models.User) {
	// Account fields
	tA.Name = fA.Name
	tA.UserID = Itoa(fU.UserID)
	tA.ID = fA.ID
}
```
Nothing is returned. You are confusing the program input (which is similar to goverter), with the program output.

**Comment by switchupcb** [comment]
@tooolbox I have updated the README prior to **v0.3**. Let me know if any other changes should be made.
